src/IonEventStream.ts

```
@@ -0,0 +1,410 @@
+                } else if(tempEvent.eventType === IonEventType.CONTAINER_END) {
+                    this.closeContainer(currentContainer.pop(), currentContainerIndex.pop());
+                } else if(tempEvent.eventType === IonEventType.SCALAR || tempEvent.eventType === IonEventType.STREAM_END) {
+                    this.eventStream.push(tempEvent);
+                } else {
+                    throw new Error('Unexpected eventType: ' + tempEvent.eventType);
+                }
+                this.reader.stepOut();
+            }
+    }
+
+    private marshalEvent() : IonEvent {
+        let currentEvent : Map<string, any> = new Map<string, any>();
+        let tid : IonType;
+        for(tid = this.reader.next(); tid !== undefined; tid = this.reader.next()) {
+
+            let fieldName = this.reader.fieldName();
+            if (currentEvent.has(fieldName)) {
+                throw new Error('Repeated event field: ' + fieldName);
+            }
+            switch (fieldName) {
+                case 'event_type' : {
+                    currentEvent.set(fieldName, this.reader.stringValue());
+                    break;
+                }
+
+                case 'ion_type' : {
+                    currentEvent.set(fieldName, this.parseIonType());
+                    break;
+                }
+
+                case 'field_name' : {
+                    currentEvent.set(fieldName, this.reader.stringValue());
+                    break;
+                }
+
+                case 'annotations' : {
+                    currentEvent.set(fieldName, this.parseAnnotations());
+                    break;
+                }
+
+                case 'value_text' : {
+                    let tempString : string = this.reader.stringValue();
+                    let tempReader : Reader = makeReader(tempString, undefined);
+                    tempReader.next();
+                    currentEvent.set('isNull', tempReader.isNull());
```

What's this?

---

how the reader conveys if a value is null, you have to explicitly check for null and the type.

---

Can that be conveyed by the absence of a value?

---

Please give an example.

---

read fails and it shows up as a null value

---

But if read fails, the test should already have failed before reaching this point.

---

yes, should have being the key word

---

Let me rephrase that, then. The test *will* have already failed before reaching this point. If it doesn't do that, then the test harness isn't implemented correctly. The test harness doesn't test itself.

---

so reader.stringvalue returns undefined when parsing a misprinted null value but I print out the correct null value because I interpreted an undefined value as null.

---

The reader should fail if it parses an invalid null value, like `null.foobar`. If the reader returning undefined is synonymous with failure, then fail as soon as a reader returns undefined. There's no opportunity to incorrectly re-write a valid null, because the test has already failed.

---

thats the goal